### PR TITLE
trim long annotation value length span

### PR DESCRIPTION
**Pull request by djia-vm-wf**

### CODE REVIEW

File: java-lib/src/test/java/com/wavefront/data/ValidationTest.java [hunk 1]
```diff
@@ -441,18 +441,13 @@ public void testInvalidSpansWithValidationConfig() {
...
     }
 
-    // span annotation value too long: WF-433
+    // accept long span annotation value
     span = getValidSpan();
     span.getAnnotations().add(new Annotation("k", "v23456789012345"));
     Validation.validateSpan(span, config);
     span = getValidSpan();
     span.getAnnotations().add(new Annotation("k", "v234567890123456"));
-    try {
```
Comment (sushantdewan123) [inline java-lib/src/test/java/com/wavefront/data/ValidationTest.java]: apart from deleting current test case, please add new test case for the code that you have changed.

File: java-lib/src/main/java/com/wavefront/data/Validation.java [hunk 2]
```diff
@@ -38,6 +40,10 @@
 
   private final static LoadingCache<String, Counter> ERROR_COUNTERS = Caffeine.newBuilder().
       build(x -> Metrics.newCounter(new MetricName("point", "", x)));
+  private final static LoadingCache<String, Counter> TRIM_COUNTERS = Caffeine.newBuilder().
+      build(x -> Metrics.newCounter(new MetricName("point", "", x)));
```
Comment (basilisk487) [inline java-lib/src/main/java/com/wavefront/data/Validation.java]: Second cache is not needed. Reuse the same one, rename it to "COUNTERS" if you don't like ERROR_COUNTERS.

File: java-lib/src/main/java/com/wavefront/data/Validation.java [hunk 3]
```diff
@@ -203,9 +209,15 @@ public static void validateSpan(Span span, @Nullable ValidationConfiguration con
...
-          throw new IllegalArgumentException("WF-433: Span annotation value is too long (" + tagV.length() +
-              " characters, max: " + config.getAnnotationsValueLengthLimit() + "): " + tagV);
+          if (blockedLoggingRateLimiter.tryAcquire()){
+            log.warning("[" + span.getCustomer() + "] Span trimmed due to tagV length: " +
+                annotation.getKey() + " limit for: " + span.getCustomer() + " is: " +
+                config.getAnnotationsValueLengthLimit() + ", found: " + annotation.getValue().length()
+                + ", span: " + span);
+            // trim the tag value to the allowed limit
+            annotation.setValue(annotation.getValue().substring(0, config.getSpanAnnotationsValueLengthLimit()));
+            TRIM_COUNTERS.get("spanAnnotationValueTooLong").inc();
```
Comment (basilisk487) [inline java-lib/src/main/java/com/wavefront/data/Validation.java]: spanAnnotationValueTruncated

File: java-lib/src/main/java/com/wavefront/data/Validation.java [hunk 4]
```diff
@@ -203,9 +209,15 @@ public static void validateSpan(Span span, @Nullable ValidationConfiguration con
...
-          ERROR_COUNTERS.get("spanAnnotationValueTooLong").inc();
-          throw new IllegalArgumentException("WF-433: Span annotation value is too long (" + tagV.length() +
-              " characters, max: " + config.getAnnotationsValueLengthLimit() + "): " + tagV);
+          if (blockedLoggingRateLimiter.tryAcquire()){
+            log.warning("[" + span.getCustomer() + "] Span trimmed due to tagV length: " +
+                annotation.getKey() + " limit for: " + span.getCustomer() + " is: " +
+                config.getAnnotationsValueLengthLimit() + ", found: " + annotation.getValue().length()
+                + ", span: " + span);
+            // trim the tag value to the allowed limit
+            annotation.setValue(annotation.getValue().substring(0, config.getSpanAnnotationsValueLengthLimit()));
```
Comment (basilisk487) [inline java-lib/src/main/java/com/wavefront/data/Validation.java]: Critical: this should not be inside `if(tryAcquire)` block, we want to fix every span, not just one per second

**Pull request merged**